Project1/PartyBase.cpp

```diff
@@ -0,0 +1,73 @@
+#include "PartyBase.h"
+
+PartyBase::PartyBase() :
+	PartyBase("~NULL PARTY~", std::vector<UnitData>())
+{}
+
+PartyBase::PartyBase(std::string name) :
+	PartyBase(name, std::vector<UnitData>())
+{}
+
+PartyBase::PartyBase(std::string name, std::vector<UnitData> unit_data) :
+	_party_name(std::move(name)) {
+	for (UnitData& data : unit_data) {
+		insertUnit(data);
+	}
+}
+PartyBase::PartyBase(PartyData data) :
+	PartyBase(data.name, data.unit_data)
+{}
+
+PartyBase::const_iterator PartyBase::getPosition(const Unit & unit) const {
+	return std::find_if(cbegin(), cend(), [&](const UnitPtr& a) -> bool { return a.get() == &unit; });
+}
+
+PartyBase::iterator PartyBase::getPosition(const Unit & unit) {
+	return std::find_if(begin(), end(), [&](const UnitPtr& a) -> bool { return a.get() == &unit; });
+}
+
+Party& PartyBase::getParty(Passkey<Map> key) {
+	return *reinterpret_cast<Party*>(this);
+}
+
+void PartyBase::insertUnit(UnitData unit) {
+	_units.emplace_back(std::make_unique<Unit>(unit));
+	_units.back()->_party = this;
+}
+
+bool PartyBase::hasUnit(const Unit& unit) const {
+	return (getPosition(unit) != cend());
+}
+
+void PartyBase::insertUnit(UnitPtr& unit) {
+	Expects(!hasUnit(*unit) && unit != nullptr);
```

Comment:
you cant have the unit since you're given an owning pointer to it

Edit: it should expect that unit->_party == nullptr, because either it was just created or another party dropped it

Reply:
Insert currently steals Units in another party, should I change that?